File: src/game/inventory/favorites.ts
import type { EquippedStack, InventoryStack, PlayerSave } from '../save/types'

export function isFavoriteStack(
  stack: InventoryStack | EquippedStack | null | undefined,
): boolean {
  return stack?.favorite === true
}

/** Favorites first; otherwise preserve relative order (stable sort). */
export function sortInventoryFavoritesFirst(save: PlayerSave): PlayerSave {
  const inventory = save.inventory
    .map((stack, index) => ({ stack, index }))
    .sort((a, b) => {
      const af = isFavoriteStack(a.stack) ? 0 : 1
      const bf = isFavoriteStack(b.stack) ? 0 : 1
      if (af !== bf) return af - bf
      return a.index - b.index
    })
    .map((entry) => entry.stack)
  return { ...save, inventory }
}

export function toggleInventoryFavorite(
  save: PlayerSave,
  index: number,
): PlayerSave | null {
  const stack = save.inventory[index]
  if (!stack) return null
  const inventory = save.inventory.map((entry, i) => {
    if (i !== index) return entry
    if (isFavoriteStack(entry)) {
      const { favorite: _removed, ...rest } = entry
      return rest
    }
    return { ...entry, favorite: true }
  })
  return sortInventoryFavoritesFirst({ ...save, inventory })
}

/** Toggles the favorite flag on a worn stack, or null when the slot is empty. */
export function toggleEquippedFavorite(save: PlayerSave, slotId: string): PlayerSave | null {
  const stack = save.equipment.slots[slotId]
  if (!stack) return null
  const next = isFavoriteStack(stack)
    ? (() => {
        const { favorite: _removed, ...rest } = stack
        return rest
      })()
    : { ...stack, favorite: true }
  return {
    ...save,
    equipment: {
      ...save.equipment,
      slots: { ...save.equipment.slots, [slotId]: next },
    },
  }
}

File: packages/ik_rules/lib/src/inventory/favorites.dart
import '../save/generated/save_models.dart';

bool isFavoriteStack(InventoryStack? stack) => stack?.favorite == true;

bool isFavoriteEquipped(EquippedStack? stack) => stack?.favorite == true;

/// Favorites first, otherwise original order.
///
/// The comparator carries the original index, so this does not depend on the
/// sort being stable the way the JavaScript version does.
PlayerSave sortInventoryFavoritesFirst(PlayerSave save) {
  final indexed = save.inventory.indexed.toList()
    ..sort((a, b) {
      final aFavorite = isFavoriteStack(a.$2) ? 0 : 1;
      final bFavorite = isFavoriteStack(b.$2) ? 0 : 1;
      if (aFavorite != bFavorite) return aFavorite - bFavorite;
      return a.$1 - b.$1;
    });
  return save.copyWith(inventory: indexed.map((entry) => entry.$2).toList());
}

/// Toggles the favorite flag on a bag stack, or null when the index is empty.
///
/// Unfavoriting drops the field entirely rather than storing `false`, matching
/// the TypeScript version so saved JSON stays identical.
PlayerSave? toggleInventoryFavorite(PlayerSave save, int index) {
  if (index < 0 || index >= save.inventory.length) return null;
  final inventory = save.inventory.indexed.map((entry) {
    if (entry.$1 != index) return entry.$2;
    final stack = entry.$2;
    return isFavoriteStack(stack) ? stack.copyWith(favorite: null) : stack.copyWith(favorite: true);
  }).toList();
  return sortInventoryFavoritesFirst(save.copyWith(inventory: inventory));
}
